osu.Game/Graphics/UserInterface/OsuTextBox.cs

```diff
@@ -59,5 +63,47 @@ protected override void OnFocusLost(FocusLostEvent e)
         }
 
         protected override Drawable GetDrawableCharacter(char c) => new OsuSpriteText { Text = c.ToString(), Font = OsuFont.GetFont(size: CalculatedTextSize) };
+
+        protected override Caret CreateCaret() => new BeatCaret
+        {
+            CaretWidth = CaretWidth,
+            SelectionColour = SelectionColour,
+        };
+
+        private class BeatCaret : BasicCaret
+        {
+            private bool hasSelection;
+
+            public BeatCaret()
+            {
+                AddInternal(new CaretBeatSyncedContainer(this));
```

Comment:
not too keen on passing self into child. instead of this, i'd propose making this caret derive `Caret` not `BasicCaret`, moving the logic we use from the basic implementation across and in the process putting the box inside the beat synced portion (so no passing of `this` is required).